HPCC-19831 Store DESDL definitions as blob to preserve element order

asselitx opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested memory usage in valgrind, and confirmed correct element ordering for newly-published services. Confirmed that definitions published using previous PTree storage method are supported. Exercised publish and get commands from the command-line and ECLWatch UI, using both supported WS versions.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-19831 Preserve element order of ESDL definitions.

 - Store defn in dali as string.
 - Maintain backward compatibility for definitions already stored as PTree.
 - Modify CEsdlSDSStore::addDefinition signature to accept defn as char* to
   avoid repeated changes to/from PTree.
 - Remove IEsdlStore API that returns definition as PTree to eliminate
   possibility of mis-ordered elements in definition.
 - Add API that gives access to definition metadata and service/
   method lists. Rework ws_esdlconfigservice to use new API.
 - Establish consistency in how the esdlConfigService returns Service and
   Method lists across the PublishESDLDefinition and GetESDLDefinition
   methods.
 - Fix memory leaks.

Code review changes include :

 - Change tag name of ESDL Definition string content element in dali from
   to 'content'.
 - Patch CEsdlSDSStore::isMethodDefined to accept new ESDL Definition
   content element.
 - Fix bug in CEsdlDefinitionInfo::queryMethods when service not found
 - Add null checks in a few places.
 - Rename methods and change prototypes to adhere to conventions.
 - Make several small style and formatting changes.
 - Use default member initilization in CEsdlDefinitionInfo.

Signed-off-by: Terrence Asselin <terrence.asselin@lexisnexisrisk.com>

asselitx commented:
@afishbeck and @timothyklemm could you please review?

HPCCSmoketest commented:
Process of PR-12475, label: hpcc-19831 starts now.
The reason of this test is: Did not build yet.
Commit ID: e0ce70ccf6f538e10ebc67d6658f409132a4b9d2
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e0ce70ccf6f538e10ebc67d6658f409132a4b9d2
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |111 |111 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |827 |827 |0 |174 sec  (00:02:54) |
|test (thor) |752 |752 |0 |616 sec  (00:10:16) |
|test (roxie) |903 |903 |0 |222 sec  (00:03:42) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   36 sec (00:00:36) |  211 sec (00:03:31) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1324 sec (00:22:04) |   18 sec (00:00:18) | 1610 sec (00:26:50) |

HPCCSmoketest commented:
Process of PR-12475, label: hpcc-19831 starts now.
The reason of this test is: New commit.
Commit ID: 2a804cb45fe7ebc2094837e7cff4dd4cf4dface5
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 2a804cb45fe7ebc2094837e7cff4dd4cf4dface5
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |111 |111 |0 |0 |0 |33 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |827 |827 |0 |190 sec  (00:03:10) |
|test (thor) |752 |752 |0 |582 sec  (00:09:42) |
|test (roxie) |903 |903 |0 |231 sec  (00:03:51) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  214 sec (00:03:34) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1288 sec (00:21:28) |   19 sec (00:00:19) | 1574 sec (00:26:14) |

timothyklemm commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -307,74 +373,44 @@ bool CWsESDLConfigEx::onPublishESDLDefinition(IEspContext &context, IEspPublishE
+                        // starting with the <Definition> element as root. However, since it's returned
+                        // in the SOAP element named <Interface>, I think the intention may have been
+                        // to return with the root <esxdl>
+                        this->wrapWithDefinitionElement(defInfo, definitionxml);
```
Let's discuss...

asselitx replied:
As per our discussion, we'll leave this as-is. It could be considered important for the GetESDLDefinition SOAP method to return a *full* DESDL Definition that could, for example, be used as input to publish that definition on another deployment. If required in the future we may consider options to the method and an updated interface that would provide the definition in a different format (eg stripped of the <Definition> element). Additionally, any change to this interface would need to be versioned because of ECLWatch's use.

timothyklemm commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -307,74 +373,44 @@ bool CWsESDLConfigEx::onPublishESDLDefinition(IEspContext &context, IEspPublishE
-                            msg.append("\nCould not fetch available methods");
+                        catch (...)
+                        {
+                            msg.append("\nEncountered error while parsing fetching available methods");
```
Grammar check

asselitx replied:
fixed

timothyklemm commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -1504,22 +1553,16 @@ bool CWsESDLConfigEx::onGetESDLDefinition(IEspContext &context, IEspGetESDLDefin
-            else
-                message.setf("Successfully fetched ESDL Defintion: %s from Dali.", id.str());
+            respcode = -1;
+            message.append("\nDefinition appears to be empty!");
```
Length of 0 is the definition of empty.

asselitx replied:
You mean this warning message should be more assertive by stating that the definition *is* empty?

timothyklemm replied:
Yes. While you're here, include the ID in the message to make debugging easier.

asselitx replied:
Ok done.

timothyklemm commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -307,74 +373,44 @@ bool CWsESDLConfigEx::onPublishESDLDefinition(IEspContext &context, IEspPublishE
+                        // starting with the <Definition> element as root. However, since it's returned
+                        // in the SOAP element named <Interface>, I think the intention may have been
+                        // to return with the root <esxdl>
+                        this->wrapWithDefinitionElement(defInfo, definitionxml);
```
Looks odd to use 'this->' here (and five other places) considering how many self references aren't prefixed.

asselitx replied:
force of habit- fixed to match style in file.

timothyklemm commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -1555,80 +1605,37 @@ bool CWsESDLConfigEx::onGetESDLDefinition(IEspContext &context, IEspGetESDLDefin
                 catch (...)
                 {
-                    message.append("\nEncountered error while parsing fetching EsdlServices");
+                    message.append("\nEncountered error while parsing fetching available methods");
```
Now I see why you need a grammar check.

asselitx replied:
fixed

HPCCSmoketest commented:
Process of PR-12475, label: hpcc-19831 starts now.
The reason of this test is: New commit.
Commit ID: a7e2e2f53a4d793ccbacaf5cee0649731ad15db1
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a7e2e2f53a4d793ccbacaf5cee0649731ad15db1
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |17 sec  (00:00:17) |
|test (hthor) |831 |831 |0 |184 sec  (00:03:04) |
|test (thor) |756 |756 |0 |609 sec  (00:10:09) |
|test (roxie) |907 |907 |0 |223 sec  (00:03:43) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  211 sec (00:03:31) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1319 sec (00:21:59) |   20 sec (00:00:20) | 1602 sec (00:26:42) |

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -85,50 +157,108 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
-            if (!conn)
-                throw MakeStringException(-1, "Unable to connect to ESDL Service definition information in dali '%s'", ESDL_DEFS_ROOT_PATH);
+            unsigned latestSeq = fetchLatestSeqForDefinitionName(definitionId);
+            targetId.appendf(".%d", latestSeq);
```
Fixed

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -172,10 +306,13 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
         }
 
-        xpath.setf("%s[@id='%s.%d'][1]/esxdl", ESDL_DEF_ENTRY, definitionName, latestSeq);
+        xpath.setf("%s[@id='%s.%d'][1]", ESDL_DEF_ENTRY, definitionName, latestSeq);
```
Yes- using method now instead.

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -320,6 +458,15 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
         else
             definitionInfo->setProp("@created",dt.getString(str).str());
 
+#ifdef ESDL_STORAGE_IS_STRING
```
It was helpful during testing and I thought it could possibly be a option for a custom compilation of the platform if a user didn't want to use the new style. If you think it should be streamlined to unconditionally use the new style I'm fine with that.

timothyklemm replied:
Considering that the purpose of this is to ensure order, and choosing not to use this guarantees that order is not maintained, I'm not sure it's something we should encourage. If there is a need for doing it, maybe the name could be changed to something that reflects the consequences of doing it.

When it comes to ensuring element order of a binding definition, I think bypassing this type of logic should not be supported. Reordering script elements could completely alter the script outcome.

asselitx replied:
OK that makes sense. I'll remove prior to my final commit.

asselitx commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -191,6 +191,76 @@ void CWsESDLConfigEx::init(IPropertyTree *cfg, const char *process, const char *
+
+        // If a service name is passed in, then only return info
+        // for that single service
+        if( !(svc && *svc) || (svc && *svc && strcasecmp(svc, serviceName)==0) )
```
Done

asselitx commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -191,6 +191,76 @@ void CWsESDLConfigEx::init(IPropertyTree *cfg, const char *process, const char *
+    }
+}
+
+void CWsESDLConfigEx::buildServiceWithMethodsResponse(IEsdlDefinitionInfo* defInfo, IArrayOf<IEspESDLService>& serviceList, const char* svc)
```
Reviewing a shared-code implementation showed that it required too many if statements, making it harder to read. Keeping this implementation as-is. Added some comments to highlight the differences between the two functions.

HPCCSmoketest commented:
Process of PR-12475, label: hpcc-19831 starts now.
The reason of this test is: New commit.
Commit ID: 6786edad5eb52277b8ff2df684285ef387136293
Estimated completion time is ~0.5 hour(s)

asselitx commented:
@timothyklemm Please review changes addressing code review comments. Also patched

- ESDL store isMethodDefined to recognize the <esxdl_blob>element
- ESDLDefinitionInfo queryMethods to fail correctly when the service is not found

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6786edad5eb52277b8ff2df684285ef387136293
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |835 |835 |0 |178 sec  (00:02:58) |
|test (thor) |759 |759 |0 |619 sec  (00:10:19) |
|test (roxie) |911 |911 |0 |224 sec  (00:03:44) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  213 sec (00:03:33) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1325 sec (00:22:05) |   19 sec (00:00:19) | 1607 sec (00:26:47) |

timothyklemm commented:
@afishbeck I've reviewed this, so feel free to add your two cents.

afishbeck commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -191,6 +191,82 @@ void CWsESDLConfigEx::init(IPropertyTree *cfg, const char *process, const char *
+        // for that single service
+        if( isEmptyString(svc) || strcasecmp(svc, serviceName)==0 )
+        {
+            Owned<IEspESDLService> esdlservice = createESDLService("","");
```
Btw, I think in cases like this ("","") may be defaulted.  You may be able to use ();

asselitx replied:
fixed

asselitx replied:
changed

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -85,50 +159,104 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+        // First get PTree for the definition
+        // In the future we may need to use an implementation that doesn't rely on
+        // PTree so we can keep the list of Services and Methods in lexical order
+        if( defn->hasProp("esxdl_blob") )
```
Could easily be convinced if you thought content was confusing for some reason.

asselitx replied:
switched to 'content'. defined static const strings for both the old 'esxdl' and the new 'content'.

afishbeck reviewed (commented):
@asselitx fairly picky and some trivial comments, but now is the time to get the interfaces right, and to make sure we're on the same page.

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+
+    MapStringToStringArray m_serviceMethodMap;
+    StringArray m_services;
+    Owned<IProperties> m_metadata;
```
Updated

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+    ~CEsdlDefinitionInfo()
+    {
+        HashIterator iter(m_serviceMethodMap);
+        for(iter.first(); iter.isValid(); iter.next())
```
Using macro

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+        if( methods )
+            return *(methods);
+        else
+            return NULL;
```
fixed

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+    }
+
+private:
+    void addMetadata(const char* name, const char* value)
```
Changed name to 'setMetadataValue()'

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+    }
+
+    void addService(const char* service)
+    {
```
Changed name

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+    {
+        if( m_serviceMethodMap.find(service) == NULL )
+        {
+            this->addService(service);
```
No it's not needed, just a style thing I've gotten used to doing. I see it isn't used widely so I'll remove.

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+        }
+
+        StringArray* methods = *m_serviceMethodMap.getValue(service);
+        methods->append(method);
```
In short it will handle that case without error by adding a duplicate entry to the array. 

Even though it would be ill-advised to do so, it is possible to publish a definition containing two methods with the same name in one service. No errors are thrown, and this code will list both of them through the 'GetESDLDefinition' method.

afishbeck replied:
I think we should consider it an error and probably check things like that when the ESDL is published.  I really don't like the idea of two methods with the same name.

asselitx replied:
Ok I'll open a ticket to add that error check (on publication only- not also here).

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+    }
+
+    void addMethodToService(const char* service, const char* method)
+    {
```
renamed

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -85,50 +159,104 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
-            if (!conn)
-             throw MakeStringException(-1, "Unable to connect to ESDL Service definition information in dali '%s'", xpath.str());
+            const char* defXML = defn->queryProp("esxdl_blob");
+            Owned<IPropertyTree> tmpTree= createPTreeFromXMLString(defn->queryProp("esxdl_blob"), ipt_caseInsensitive|ipt_ordered);
```
done

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+        return &m_services;
+    }
+
+    virtual StringArray* queryMethods(const char* service) override
```
never mind about the &, I see you return nullptr below.

asselitx replied:
changed to `const StringArray *`

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -85,50 +159,104 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+            if( esxdlTree != nullptr )
+            {
+                readServiceMethodInfo(defInfo, esxdlTree);
+            } else {
```
updated style to match at this location and one or two others in file.

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -40,6 +40,80 @@ static const char* ESDL_CHANGE_PATH="/ESDL/Subscription/Change";
+        }
+    }
+
+    virtual IProperties* queryMetadata() override
```
changed to `const IProperties &`

asselitx replied:
yes, changed

asselitx commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -28,16 +28,23 @@
 
+interface IEsdlDefinitionInfo : public IInterface
+{
+    virtual IProperties* queryMetadata() = 0;
```
changed both queryMetadata and queryServices to return const references, and renamed to getMetadata and getServices to avoid confusion with linked object interfaces

asselitx replied:
changed to const

asselitx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -139,12 +267,16 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
 
-        toXML(conn->queryRoot(), esxdl, 0, 0);
+        IPropertyTree* defn = conn->queryRoot();
+        if( defn->hasProp("esxdl_blob"))
```
same as above

asselitx replied:
changed as above

HPCCSmoketest commented:
Process of PR-12475, label: hpcc-19831 starts now.
The reason of this test is: New commit.
Commit ID: e7dc88829dcfd6728acd7e86463547d27d336d42
Estimated completion time is ~0.5 hour(s)

asselitx commented:
@afishbeck I've checked in the changes you requested, please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e7dc88829dcfd6728acd7e86463547d27d336d42
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |847 |847 |0 |180 sec  (00:03:00) |
|test (thor) |770 |770 |0 |696 sec  (00:11:36) |
|test (roxie) |923 |923 |0 |217 sec  (00:03:37) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  254 sec (00:04:14) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1403 sec (00:23:23) |   20 sec (00:00:20) | 1729 sec (00:28:49) |

afishbeck reviewed (approved):
@asselitx looks good.

afishbeck commented:
@asselitx please squash into single commit then ask Richard to merge.

asselitx commented:
@richardkchapman could you please merge? Tony has approved changes.

HPCCSmoketest commented:
Process of PR-12475, label: hpcc-19831 starts now.
The reason of this test is: New commit.
Commit ID: c185ad900eee71973543a65daba639984a090ce1
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: c185ad900eee71973543a65daba639984a090ce1
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |26 sec  (00:00:26) |
|test (hthor) |847 |847 |0 |190 sec  (00:03:10) |
|test (thor) |770 |770 |0 |699 sec  (00:11:39) |
|test (roxie) |923 |923 |0 |256 sec  (00:04:16) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  258 sec (00:04:18) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1476 sec (00:24:36) |   20 sec (00:00:20) | 1807 sec (00:30:07) |
